client-v2/src/shared/util/shared.ts

```diff
@@ -26,38 +26,78 @@ const createGroup = (
 
 const getUUID = <T extends { uuid: string }>({ uuid }: T) => uuid;
 
-// this is horrible but is the only way to do it!
-const groupByIndex = (groups: Group[], index: number): Group | undefined =>
-  groups.find(g => parseInt(g.id || '0', 10) === index);
+const getGroupIndex = (id: string | null): number => parseInt(id || '0', 10);
 
 const getAllArticleFragments = (groups: Group[]) =>
   groups.reduce(
     (acc, { articleFragments }) => [...acc, ...articleFragments],
     [] as string[]
   );
 
-const addEmptyGroupsFromCollectionConfigForStage = (
+const configGroupIndexExistsInGroups = (
+  groupsToSearch: Group[],
+  index: number
+): boolean =>
+  groupsToSearch.some(group => {
+    if (group.id) {
+      return parseInt(group.id, 10) === index;
+    }
+    return index === 0;
+  });
+
+const addGroupsForStage = (
   groupIds: string[],
   entities: { [id: string]: Group },
   collectionConfig: CollectionConfig
 ) => {
   const groups = groupIds.map(id => entities[id]);
-  const addedGroups = collectionConfig.groups
-    ? collectionConfig.groups.map((name, i) => {
-        const existingGroup = groupByIndex(groups, i);
+  let groupsWithNames = groups.map(group => {
```

Comment:
Super minor: do we need a `let` here? As far as I can see the only reassignment occurs on 79, and this reassignment could be a `push` as it only occurs when `groupsWithNames` is empty.